src/axom/primal/geometry/BezierPatch.hpp

```diff
@@ -0,0 +1,875 @@
+              for(int k = 0; k <= end; ++k)
+              {
+                dCarray[k] = lerp(dCarray[k], dCarray[k + 1], uv);
+                dWarray[k] = lerp(dWarray[k], dWarray[k + 1], uv);
+              }
+            }
+            c[p][i] = dCarray[0] / dWarray[0];
+            c.setWeight(p, dWarray[0]);
+          }
+      }
+    }
+    else
+    {
+      if(axis == 0)  // Keeping a fixed value of u
+      {
+        c.setOrder(ord_v);
+        axom::Array<T> dCarray(ord_u + 1);  // Temp array
+
+        // Run de Casteljau algorithm on each row of control nodes and each dimension
+        for(int q = 0; q <= ord_v; ++q)
+          for(int i = 0; i < NDIMS; ++i)
+          {
+            for(int p = 0; p <= ord_u; ++p)
+              dCarray[p] = m_controlPoints(p, q)[i];
+
+            for(int p = 1; p <= ord_u; ++p)
+            {
+              const int end = ord_u - p;
+              for(int k = 0; k <= end; ++k)
+                dCarray[k] = lerp(dCarray[k], dCarray[k + 1], uv);
+            }
+            c[q][i] = dCarray[0];
+          }
+      }
+      // Run de Casteljau algorithm on each column of control nodes and each dimension
+      else  // Keeping a fixed value of v
+      {
+        c.setOrder(ord_u);
+        axom::Array<T> dCarray(ord_v + 1);  // Temp array
+
+        for(int p = 0; p <= ord_u; ++p)
+          for(int i = 0; i < NDIMS; ++i)
+          {
+            for(int q = 0; q <= ord_v; ++q)
+              dCarray[q] = m_controlPoints(p, q)[i];
+
+            for(int q = 1; q <= ord_v; ++q)
+            {
+              const int end = ord_v - q;
+              for(int k = 0; k <= end; ++k)
+              {
+                dCarray[k] = lerp(dCarray[k], dCarray[k + 1], uv);
+              }
+            }
+            c[p][i] = dCarray[0];
+          }
+      }
+    }
+
+    return c;
+  }
+
+  /*!
+   * \brief Evaluates a Bezier patch at a particular parameter value (\a u, \a v)
+   *
+   * \param [in] u parameter value at which to evaluate on the first axis
+   * \param [in] v parameter value at which to evaluate on the second axis
+   * \return p the value of the Bezier patch at (u, v)
+   *
+   * \note We typically evaluate the patch at \a u and \a v between 0 and 1
+   */
+  PointType evaluate(T u, T v) const
+  {
+    if(getOrder_u() >= getOrder_v())
+      return isocurve(u, 0).evaluate(v);
+    else
+      return isocurve(v, 1).evaluate(u);
+  }
+
+  /*!
+   * \brief Computes a tangent of a Bezier patch at a particular parameter value (\a u, \a v) along \a axis
+   *
+   * \param [in] u parameter value at which to evaluate on the first axis
+   * \param [in] v parameter value at which to evaluate on the second axis
+   * \param [in] axis orientation of vector. 0 for fixed u, 1 for fixed v
+   * \return v a tangent vector of the Bezier patch at (u, v)
+   *
+   * \note We typically find the tangent of the patch at \a u and \a v between 0 and 1
+   */
+  VectorType dt(T u, T v, int axis) const
```

Comment:
Would we be interested in directional derivatives? If so, this might take an angle instead of an axis as the direction.

It'd be more complicated to implement though (given how you set this up in terms of the isocurves), so perhaps we should wait until there's a need.

Reply:
I don't think it's that much more complicated to implement, as you just compute both axis-aligned tangents and do a linear combination of those, but I think it makes the interface more complicated. Because the two axis-aligned ones are (probably) the cases most likely to be used, and they require half the computation time of an arbitrary direction, so you wouldn't want to compute every `dt(axis = 0)` as a `dt( theta = 0 )` naively. It's definitely something that could be used down the road, but I think it would be okay to push that question down the line until we know how the tangents are being used.